src/backend/access/heap/heapam_visibility.c

```diff
@@ -1143,6 +1190,178 @@ HeapTupleSatisfiesMVCC(HeapTuple htup, Snapshot snapshot,
 	return false;
 }
 
+/*
+ * RemoteHeapTupleSatisfiesMVCC
+ *		True iff heap tuple on a remote relation is valid for the given MVCC snapshot.
+ *
+ * This function is structurally similar to the HeapTupleSatisfiesMVCC function with new if branches
+ * for when xmin or xmax is determined to come from a remote region. If xmin (or xmax) has been created
+ * locally, this function uses the same code as HeapTupleSatisfiesMVCC, otherwise, it uses
+ * RemoteTransactionIdDidCommit, the remote variation of TransactionIdDidCommit, to determine whether
+ * an xid is committed or not.
+ */
+static bool
+RemoteHeapTupleSatisfiesMVCC(int region, HeapTuple htup, Snapshot snapshot, Buffer buffer)
+{
+	HeapTupleHeader tuple = htup->t_data;
+
+	Assert(ItemPointerIsValid(&htup->t_self));
+	Assert(htup->t_tableOid != InvalidOid);
+
+	if (!HeapTupleHeaderXminCommitted(tuple))
+	{
+		if (HeapTupleHeaderXminInvalid(tuple))
+			return false;
+		
+		/*
+		 * Use the same code path as normal postgres if xmin is local (implying xmax is also local)
+		 */
+		if (HeapTupleHeaderIsXminLocal(tuple))
+		{
+			if (TransactionIdIsCurrentTransactionId(HeapTupleHeaderGetRawXmin(tuple)))
+			{
+				if (HeapTupleHeaderGetCmin(tuple) >= snapshot->curcid)
+					return false;	/* inserted after scan started */
+
+				if (tuple->t_infomask & HEAP_XMAX_INVALID)	/* xid invalid */
+					return true;
+
+				if (HEAP_XMAX_IS_LOCKED_ONLY(tuple->t_infomask))
+					return true;
+
+				/* The tuple is created by and local to the current xact so must not have multixact */
+				Assert(!(tuple->t_infomask & HEAP_XMAX_IS_MULTI));
+
+				/* Only current transaction can delete this tuple */
+				Assert(TransactionIdIsCurrentTransactionId(HeapTupleHeaderGetRawXmax(tuple)));
+
+				if (HeapTupleHeaderGetCmax(tuple) >= snapshot->curcid)
+					return true; /* updated after scan started */
+				else
+					return false;	/* updated before scan started */
+			}
+			else if (TransactionIdDidCommit(HeapTupleHeaderGetRawXmin(tuple)))
+				SetHintBits(tuple, buffer, HEAP_XMIN_COMMITTED,
+							HeapTupleHeaderGetRawXmin(tuple));
+			else
+			{
+				/* it must have aborted or crashed */
```

Comment:
nit: You can't really crash in a local xact and continue to execute next subxact. So maybe just "it must have aborted/rolledback"

Reply:
This part was copied-pasted from HeapTupleSatisfiesMVCC